File: icecube-tau-decays/pythia_tau_decay.py
import pythia8
import numpy as np
import pandas as pd
import argparse


def main():
    # Get file name to read from and to write to from command line
    parser = argparse.ArgumentParser(description="Simulate tau decays using Pythia8")
    parser.add_argument("input_file", help="CSV file containing particle information")
    parser.add_argument("output_file", help="CSV file to write decay information to")
    parser.add_argument("-p", "--polarization",
                        dest="p",
                        type=float,
                        help="Polarization magnitude z direction. A float is possible, but the polarization can only be fully parallel It might be that -1, 0, or 1 are the only possible values, or maybe floats are possible", 
                        default=0)
    
    args = parser.parse_args()

    input_filename = args.input_file
    output_filename = args.output_file
    polarization = args.p
    
    print(polarization)

    tau_pdg_id = 15  # PDG ID for tau lepton

    # Load particles simulated by IceCube
    particle_info = pd.read_csv(input_filename)
    
    # Initialize Pythia
    pythia = pythia8.Pythia()
    # pythia.readString("ProcessLevel:resonanceDecays=on")
    # this is the trick to make Pythia8 work as "decayer"
    pythia.readString("ProcessLevel:all = off")
    pythia.readString("ProcessLevel:resonanceDecays=on")
    
    # Set tau polarization. See https://pythia.org/latest-manual/ParticleDecays.html#section2
    pythia.readString(f"TauDecays:mode = 3")
    pythia.readString(f"TauDecays:tauPolarization = {polarization}")
    # shut off Pythia8 (default) verbosity
    # fDecayer->readString("Init:showAllSettings=false");
    # fDecayer->readString("Init:showChangedSettings=false");
    # fDecayer->readString("Init:showAllParticleData=false");
    # fDecayer->readString("Init:showChangedParticleData=false");
    
    # specify how many Py8 events to print out, at either level
    # in this particular case print out a maximum of 10 events
    # fDecayer->readString("Next:numberShowProcess = 0" );
    # fDecayer->readString("Next:numberShowEvent = 10");
    
    # Disable pion decay
    pythia.readString("111:onMode = off")
    # Disable some other decays. Not sure if these really help. Maybe I should do this for other hadrons, e.g. pi- too
    pythia.readString("310:onMode = off")
    pythia.readString("221:onMode = off")

    pythia.init()
    with open(output_filename, "w") as f:
        f.write("event_num,pdg,E,px,py,pz\n")

    for event_num, particles in particle_info.groupby("event_num"):
        tau = particles[particles["pdg"] == tau_pdg_id].iloc[0, :]

        for particle in particles.iloc[:3, :].itertuples():
            with open(output_filename, "a") as f:
                f.write(f"{particle.event_num:.0f},{particle.pdg:.0f},{particle.E},{particle.px},{particle.py},{particle.pz}\n")

        # Initialize the event
        pythia.event.reset()

        # Calculate tau mass from 4-momentum
        tau_mass = np.sqrt(tau.E**2 - (tau.px**2 + tau.py**2 + tau.pz**2))
        # Add tau lepton to the event
        pythia.event.append(
            tau_pdg_id, 1, 0, 0, 
            #  px      py      pz      E       m
            tau.px, tau.py, tau.pz, tau.E, tau_mass
        )
        
        # Set polarization. This seems to have the opposite effect
        # pythia.event.back().pol(polarization)

        # Generate the decay
        pythia.next()

        # Access the decay products
        event_particles = pythia.event
        
        # print(f"Event number: {tau.event_num}")
        # Print information about the decay products
        for i, particle in enumerate(event_particles):
            if not particle.isFinal():
                continue

            with open(output_filename, "a") as f:
                f.write(f"{tau.event_num:.0f},{particle.id():.0f},{particle.e()},{particle.px()},{particle.py()},{particle.pz()}\n")
        
    # End Pythia
    pythia.stat()

    # Convert the decay information to a DataFrame
    # pythia_decays_df = pd.DataFrame(pythia_decays, columns=["event_num", "pdg", "E", "px", "py", "pz"])
    # pythia_decays_df.to_csv(output_filename, index=False)


if __name__ == "__main__":
    main()

File: icecube-tau-decays/pythia_tau_decay.cc
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <map>
#include <iomanip>
#include "Pythia8/Pythia.h"

struct SimpleParticle {
    int event_num;
    int pdg;
    double E;
    double px;
    double py;
    double pz;
};

std::vector<SimpleParticle> readCSV(const std::string& filename) {
    std::ifstream file(filename);
    std::vector<SimpleParticle> particles;
    std::string line;

    // Skip header
    std::getline(file, line);

    while (std::getline(file, line)) {
        std::istringstream iss(line);
        std::string token;

        SimpleParticle particle;
    
        std::getline(iss, token, ',');
        particle.event_num = std::stoi(token);

        std::getline(iss, token, ',');
        particle.pdg = std::stoi(token);

        std::getline(iss, token, ',');
        particle.E = std::stod(token);

        std::getline(iss, token, ',');
        particle.px = std::stod(token);

        std::getline(iss, token, ',');
        particle.py = std::stod(token);

        std::getline(iss, token);
        particle.pz = std::stod(token);

        particles.push_back(particle);
    }

    return particles;
}


int main(int argc, char* argv[]) {
    if (argc != 4) {
        std::cerr << "Usage: " << argv[0] << " input_file output_file polarization" << std::endl;
        return 1;
    }

    const std::string input_filename = argv[1];
    const std::string output_filename = argv[2];

    const double polarization = std::stod(argv[3]);

    // Read particles from CSV
    std::vector<SimpleParticle> particles = readCSV(input_filename);

    // Initialize Pythia8
    Pythia8::Pythia pythia;

    // Set tau decays mode and polarization
    pythia.readString("ProcessLevel:all = off");
    pythia.readString("ProcessLevel:resonanceDecays=on");
    pythia.readString("TauDecays:mode = 3");
    pythia.readString("TauDecays:tauPolarization = " + std::to_string(polarization));

    // Disable some decays
    pythia.readString("111:onMode = off");
    pythia.readString("310:onMode = off");
    pythia.readString("221:onMode = off");

    pythia.init();

    // Open output file in append mode
    std::ofstream outfile(output_filename);

    outfile << "event_num,pdg,E,px,py,pz\n";

    for (const auto& particle : particles) {
        // Write dummy nucleus and neutrino
        outfile << particle.event_num << "," << 1000080160 << "," 
                << 1 << "," << 0 << "," << 0 << "," << 0 << "\n";
        outfile << particle.event_num << "," << 16 << "," 
                << 1 << "," << 0 << "," << 0 << "," << 1 << "\n";
        // Write tau lepton
        outfile << particle.event_num << ","
                << particle.pdg << ","
                << std::setprecision(16) << particle.E << ","
                << std::setprecision(16) << particle.px << ","
                << std::setprecision(16) << particle.py << ","
                << std::setprecision(16) << particle.pz << "\n";

        // Reset Pythia8 event
        pythia.event.reset();

        // Calculate tau mass
        double tau_mass = std::sqrt(particle.E * particle.E - (particle.px * particle.px + particle.py * particle.py + particle.pz * particle.pz));

        // Add tau lepton to the event
        pythia.event.append(particle.pdg, 1, 0, 0, particle.px, particle.py, particle.pz, particle.E, tau_mass);

        // Generate the decay
        pythia.next();

        // Access the decay products
        for (int i = 0; i < pythia.event.size(); ++i) {
            const Pythia8::Particle& decay_product = pythia.event[i];
            if (decay_product.isFinal()) {
                outfile << particle.event_num << ","
                        << decay_product.id() << ","
                        << std::setprecision(16) << decay_product.e() << ","
                        << std::setprecision(16) << decay_product.px() << ","
                        << std::setprecision(16) << decay_product.py() << ","
                        << std::setprecision(16) << decay_product.pz() << "\n";
            }
        }
    }

    pythia.stat();

    return 0;
}
